### ISSUE TITLE
Proposal: Starlark (and others) as alternative to HCL

### ISSUE THREAD

**Issue by nikolay** [opentofu/opentofu#2297]
### OpenTofu Version

```shell
N/A
```

### The problem in your OpenTofu project

In the real world, often you need several levels of nested comprehension, which makes the code unreadable and fragile. It would be nice if OpenTofu takes the leads and mplements pluggable formats such as [Starlark](https://github.com/bazelbuild/starlark), which is relatively widely used among the DevOps community.

### Attempted Solutions

* [CDKTF](https://developer.hashicorp.com/terraform/cdktf)
* [Pulumi](https://www.pulumi.com/)
* [Terragrunt](https://terragrunt.io/)
* [Wing](https://www.winglang.io/)
* [AsCode](https://ascode.run/)

### Proposal

[AsCode](https://ascode.run/) exists as a pre-processor to HCL and it's very ergonomic, but not an official part of the ecosystem and tooling. Adding a language that's more widely adopted than HCL would boost adoption and give a significant leverage.

Dynamic generation of HCL via tools such as [Terramate](https://terramate.io/) and [Terragrunt](https://terragrunt.io/) lead to hard-to-test and catch errors due to HCL being used in templates as strings without syntax highlighting, linting, and validation.

**Comment by nikolay** [comment]
@abstractionfactory Starlark is not much different than CDKTF and CDKTF also uses imperative languages as well. I understand HCL is deeply embedded, but it could be translated into an interface from something that's both an interface and an implementation. Let's face it, we are at HashiCorp's mercy in regard to HCL - even its name carries the legacy and it's very limited in adoption due to being tied to a corporation with a not-so-stellar reputation in the open-source community. It's pathetic that HCL doesn't even have its own CLI, the formatting is tied to Terraform, Nomad, etc., and after all these years, it doesn't have a proper `jq` implementation, a linter, and a formatter that does more than indentation. To this date, there's no proper JSON-to-HCL tool that produces traditionally looking HCL! I've been hopeful that [dasel](https://daseldocs.tomwright.me/) [can bridge](https://github.com/TomWright/dasel/issues/98) that gap, but, again, it always comes down to formatting. Of course, I've tried all the tools "supporting" HCL: [qq](https://github.com/JFryy/qq), [hcledit](https://github.com/minamijoyo/hcledit) and, again, [hcledit](https://github.com/mercari/hcledit), [hclgrep](https://github.com/magodo/hclgrep) - they all lack basic ergonomics. [hclq](https://github.com/mattolenik/hclq) has also been long dead! My 16-year son who is into Rust is always joking that I'm not a software engineer as I "write HCL files all day long". I recently had to do a massive migration from Terraform Cloud, well, HCP Terraform to Atlantis, and I really had to go through the pain of having to programmatically change hundreds if not thousands of HCL files! Instead, a higher-level language comes with a proper language server, refactoring tools, etc.

Again, the issue with HCL is that it lacks adoption and why would anyone adopt it if there's [TOML](https://toml.io/en/), [Starlark](https://github.com/google/starlark-go/), [Jsonnet](https://jsonnet.org/), etc., which have wider applications and bigger communities. Well, HCL2 gave HCL a boost over TOML, but regardless, it will always stay niche, stale, subpar, neglected, lacking toolchain, support, etc.

**Comment by nikolay** [comment]
By the way, I've [added an issue](https://github.com/mikefarah/yq/issues/1844) to [yq](https://mikefarah.gitbook.io/yq/) as well!

**Comment by apparentlymart** [comment]
Putting aside for a moment the fact that OpenTofu's language runtime communicates directly with HCL's API, I want to say a little about what OpenTofu's execution model needs from any hypothetical replacement language.

- It must be possible to enumerate the top-level declarations before fully evaluating expressions inside them, so that OpenTofu can build its dependency graph and evaluate objects in the correct order.

    For example, OpenTofu needs to be able to discover that there is something equivalent to a `resource "aws_instance" "example"` block before evaluating any expressions that might include a reference to that block and before evaluating the contents of that block.

- There must be something comparable to "unknown values" that can serve as placeholders in expressions to allow OpenTofu to produce partial results during the planning phase and then re-evaluate with more information during the apply phase.

    This one is unfortunately the trickiest to match in other languages: the type system under HCL has this feature largely _because_ software like OpenTofu needs it, whereas more general-purpose languages don't tend to assume such a constraint.

    Pulumi emulates this using a promise-like model, which means that [a saved plan can fail during apply if the program behaves differently once values become "more known"](https://www.pulumi.com/blog/announcing-public-preview-update-plans/). OpenTofu's execution model instead assumes that such differences ought to be impossible at the language level and so always blames any inconsistencies on the providers generating the planned actions, since that's where the imperative code rules live in OpenTofu.

I've researched various alternatives over the years and found each one missing some part of this that makes it challenging to integrate. I investigated _Starlark in particular_ some years ago now and found it promising but unfortunately not sufficient to meet the two main requirements I listed above, although it's closer than some others: in principle it's possible to evaluate a Starlark program with nothing in scope initially and expect it to produce a program containing lambda functions in any place where a dynamic expression would appear in OpenTofu's native language, but Starlark's function type doesn't really give enough information to determine what symbols the function depends on for calculating dependency order, and as far as I can tell there's nothing analogous to "unknown values" in the language.

If we _could_ find a language that met these criteria then we could think about what it would take to integrate it into OpenTofu. One possible strategy is to implement the [`hcl.Body`](https://pkg.go.dev/github.com/hashicorp/hcl/v2#Body) and [`hcl.Expression`](https://pkg.go.dev/github.com/hashicorp/hcl/v2#Expression) interfaces[^1] in terms of the other language's evaluation functions, which is an approach I've had some partial success with in the past aside from the target languages not being able to meet all of the requirements above. This strategy has the advantage that OpenTofu Core need not know too much about it since it would access the new language through the same abstract HCL API it's already using, but of course also means that whatever language we choose has to have an API that's sufficiently compatible with HCL expectations.

Pulumi Engine was designed from the ground up to run an arbitrary language runtime concurrently with the engine and propagate data back and forth, so I tend to think that folks who want to use general-purpose programming languages to describe their infrastructure would be better served by Pulumi than by OpenTofu. OpenTofu's programming model is more like a spreadsheet, with interdependent expressions and data-flow-oriented execution, and so any alternative language we might try to support for OpenTofu would probably need to offer a similar execution model unless our intention were to totally redesign the OpenTofu runtime at the same time.

Preprocessors are somewhat easier but of course also very limited: they turn this into a metaprogramming problem rather than just a programming problem, typically forcing their users to be familiar both with the preprocessor's chosen language _and_ the OpenTofu language, and so at high risk of making things more complicated without making things any better. Pulumi's model of integrating external language runtime into the engine's event loop still causes _some_ element of "metaprogramming" in that you are writing a program to generate another program for Pulumi Engine to execute, but in Pulumi's case data can flow both ways between Pulumi Engine and the external language runtime in a similar way to how in OpenTofu data can flow bidirectionally between OpenTofu's core runtime and the HCL expression evaluator.

[^1]: Of course we could design some HCL-agnostic interfaces to use instead of these ones, but HCL's API is more-or-less designed to be just what an OpenTofu-like application needs and so any interface we'd design here would end up being very similar to what HCL's interfaces require: OpenTofu's predecessor largely imposed its requirements on HCL rather than the other way around.

**Comment by nikolay** [comment]
@abstractionfactory Either improve HCL (by forking it and rebranding it to something neutral, not OpenTofu-related, such as XCL, i.e., eXtensible Configuration Language) to be more than just a dialect of JSON with syntax for functions and make it appealing outside of the OpenTofu world so that it gets better adoption among various toolchains and IDEs, or find an alternative such as Starlark. Without Mitchell, honestly, I don't see HCL3 anytime soon. The Lua and Go interpreters via scripting providers, which you pointed out, work somewhat, but they are still unable to generate HCL. Going into the preprocessor direction is also not so great because then error reporting will point to HCL that does not exist in the source, so, just like with traditional templating, it makes debugging unpleasant.

**Comment by nikolay** [comment]
@abstractionfactory I didn't know about [KCL](https://www.kcl-lang.io/) until today, but given it's joining CNCF, that would be a better alternative then, as you (could) use one language across Terraform, Kubernetes, and beyond. In fact, it already [could be used](https://www.kcl-lang.io/docs/next/user_docs/guides/working-with-terraform/validation) (kind of) for Terraform policy validation instead of OPA/Rego.

**Comment by apparentlymart** [comment]
Thanks for letting us know about that project, @nikolay.

I have not yet had time to dig into all of the details, but from some short time reading the getting started guides and some of the reference documentation it seems like KCL is in similar vein to CUE and Jsonnet: a language oriented towards generating, transforming, or validating JSON-like or YAML-like data structures based on expressions.

As with CUE and Jsonnet, it unfortunately doesn't seem to currently be suitable as a replacement for HCL as OpenTofu uses it, since it doesn't meet the two main requirements I previously described.

It also appears to be written in Rust, which means it would not be straightforward to link it into our Go codebase without introducing CGo, a system linker, and the Rust compiler. The [KCL Go SDK](https://pkg.go.dev/kcl-lang.io/kcl-go) seems to work by using the system's dynamic linker to load the KCL engine as a shared library, but currently supports only `darwin`, `freebsd`, `linux`, and `windows` targets. Since OpenTofu is currently distributed as a single statically-linked executable, it would be tricky to switch to a model involving a separate dynamic library.

Therefore I don't really see any clear path to replacing HCL with KCL in OpenTofu. It might potentially work in a Pulumi language host, since those already run as separate processes alongside Pulumi engine and so can be written in any language, and Pulumi's bindings-generation tools could presumably produce a provider-specific module containing [schema definitions](https://www.kcl-lang.io/docs/reference/lang/spec/schema) for all of the resource types in the provider. However, I'm not sure if there's anything in this language for representing placeholders for values that are not yet known, which is a requirement for both Pulumi and OpenTofu.

However, I do agree that it seems like it can meet similar use-cases as CUE and Jsonnet, including making decisions based on `tofu show -json` output as you noted. KCL could also be used as a preprocessor for generating [OpenTofu's JSON configuration format](https://opentofu.org/docs/v1.7/language/syntax/json/), as some are already doing with Cue and Jsonnet.

**Comment by apparentlymart** [comment]
I've been pondering this issue on-and-off since it was opened, since it seemed like there was a general point to be made but it took me a while to find a plausible framing for it. I don't know if I've succeeded, but here's an attempt anyway. 😀 

I see OpenTofu, and the project it was forked from, as having a lot of similarities to at least two other established systems:

- [GNU make](https://www.gnu.org/software/make/) (and other implementations of `make`, but I'm less familiar with those) is a dependency-based process execution engine that is tightly coupled with a domain-specific programming language that is tailored to the problem of describing relationships between files and the imperative actions required to produce derived files from source files.

    The execution engine and the domain-specific language were designed together and are tightly coupled to one another. The concepts of the surface language and of the runtime implementation have considerable overlap. Any new surface language used in make would need to offer a similar set of concepts in order to be properly integrated with the execution engine.

    However, various projects have treated the make language and execution engine as an implementation detail of their higher-level model. For example, [CMake](https://cmake.org/) is a higher-level build system whose execution model is to generate a program for a lower-level execution tool to handle, and GNU make is one of the targets it supports.

    There are also other projects that aim to meet similar use-cases as GNU make does, like [Ninja](https://ninja-build.org/). Unlike make, Ninja's design prioritizes being used as an implementation detail of a build system like CMake, and so it has its own domain-specific language that is designed first for being relatively easy to generate, rather than prioritizing hand-authoring by humans. Although Ninja meets many of the same use-cases as GNU make, it replaces both the runtime engine and the surface language together, because again those two components are designed together and tightly coupled.

- [Node.js](https://nodejs.org/) is an execution engine and I/O reactor for software written in JavaScript. Despite JavaScript being a general-purpose programming language, Node.js is nonetheless relatively tightly coupled to it: it would be possible in principle to create something _like_ Node around another language runtime like Python or Lua, but arguably what resulted would not be Node because Node is defined by its relationship to JavaScript.

    Various other language projects have targeted node by "transpiling" to JavaScript. I think the most prominent example at the time I'm writing this is [TypeScript](https://www.typescriptlang.org/), but I would also place [CoffeeScript](https://coffeescript.org/) and [Emscripten](https://emscripten.org/) into this category. Similar to CMake and Make, TypeScript is treating the JavaScript language and its executor as an implementation detail, offering additional features on top of what JavaScript can offer alone.

    [Node.js has _some_ support for running TypeScript](https://nodejs.org/en/learn/typescript/run-natively), but at the time of writing it achieves it by essentially just throwing away all of the TypeScript extensions to directly lower to something the JavaScript runtime can execute. While I'm not familiar with the details of this at all, I see it as a signal of what a tight relationship there is between Node.js and the JavaScript language: TypeScript was designed from the outset to be a strict JavaScript superset such that it can always be transpiled to JavaScript, but Node.js's support for TypeScript is essentially still just a very basic transpiler rather than a direct integration of the TypeScript language.

- I don't think I need to define OpenTofu here in the same way as I attempted to define the others 😀, but much like GNU Make it has its own language that was designed together with its language runtime such that the two are each designed with the other's direct needs in mind and have considerable overlapping concepts.

    The fact that OpenTofu's language is implemented in terms of the HCL language toolkit rather than directly as a byte-stream parser tends to muddy the waters a little: it's understandable for someone outside the project to think of "HCL" and "the OpenTofu language" as synonymous, but the true situation is more like the relationship between [XML](https://en.wikipedia.org/wiki/XML) and [the Atom Syndication Format](https://en.wikipedia.org/wiki/Atom_(web_standard)): Atom is a domain-specific language defined in terms of XML's infoset, and the OpenTofu language is a domain-specific language defined in terms of HCL's infoset.

    Nonetheless, the OpenTofu language is tightly integrated with both the OpenTofu language runtime and the HCL language toolkit. None of these three items are designed to be a swappable component.

    HCL's infoset _is_ defined in such a way that it can potentially be implemented in terms of other languages, as demonstrated by [the JSON mapping to the HCL infoset](https://github.com/hashicorp/hcl/blob/main/json/spec.md). However, that only works to the extent that the new candidate language can represent the concepts in [the HCL syntax-agnostic information model](https://github.com/hashicorp/hcl/blob/main/spec.md) (which I'm abbreviating here to "infoset"), much as TypeScript code can only be executed by Node.js because it's compatible enough with JavaScript to be able to treat it as purely a syntax-mapping problem. Some time ago folks made efforts to map the XML infoset onto JSON syntax in the hope that it would make it easier to adapt languages like Atom to JSON, but those efforts failed in large part because JSON and XML are not built from a compatible set of concepts.

    The other languages we've been discussing here do not seem to provide sufficient concepts to have the HCL infoset mapped onto them, and therefore to have the OpenTofu language mapped onto them. Since OpenTofu's runtime is tightly coupled with the OpenTofu language, any other language we select _must_ be able to directly substitute for the concepts of the OpenTofu language, which are described in turn by concepts of HCL's infoset.

    Similar to what we saw with CMake and TypeScript, there are also various projects that aim to use OpenTofu or its predecessor as an implementation detail of a higher-level abstraction, such as [CDK for Terraform](https://github.com/hashicorp/terraform-cdk) or (to some extent) [Terragrunt](https://terragrunt.gruntwork.io/). [OpenTofu's alternative JSON input syntax](https://opentofu.org/docs/v1.7/language/syntax/json/) (which uses HCL's syntax-agnosticism as an implementation detail) is perhaps somewhat analogous to Ninja's offering of a language that prioritizes being generated by other software over human-authoring, although of course not all wrapping tools use that facility.

- Pulumi took a significantly different route than any of the examples I've given above: instead of co-designing a language and a runtime engine, they instead designed a runtime engine and a wire protocol (the "language host" protocol) that can be used to interact with any external plugin that implements that protocol. The Pulumi engine concepts are therefore tightly coupled with that protocol, but _not_ coupled to any particular implementation of that protocol.

    The existence of Pulumi is of course proof that OpenTofu could potentially also be designed in this way, and I expect the same could be said for a language-agnostic engine offering similar results as GNU make or Ninja. However, that generality comes at a significant complexity cost, because something that need be designed and implemented only once and tailored directly to OpenTofu's language must typically be designed and implemented separately for each target language in Pulumi.

    I don't think it will be a surprise to anyone that I prefer having a domain-specific language that's tailored to the problem space and tightly integrated with the runtime, but I also understand that not everyone shares that preference and so I am glad that Pulumi exists. I don't think it would be useful to try to transform OpenTofu to follow Pulumi's model, because Pulumi already exists and serves its constituency well. Those who want to use general-purpose programming languages to describe their infrastructure should use Pulumi to do it; that is not a situation that OpenTofu is designed to serve nor well-suited to serve directly.

---

So what can we take away from all of this? Probably lots of things, but the following are some things that _I_ take away from it:

- Domain-specific languages tend to be tightly coupled to the system whose domain they are specific to. That seems to just be the nature of the beast, since this has been true for several other DSLs before it was true for OpenTofu.

    Even _general-purpose languages_ tend to be tightly coupled with specific execution environments unless either someone makes a concerted effort to avoid that (as with Pulumi) or the execution environment is sufficently low-level that its concepts are far removed from the higher-level languages running on it, like the JVM, or an actual CPU. 😀 

- Although in principle HCL is designed to be surface-syntax-agnostic, it still has a fundamental set of concepts that any target syntax must be able to somehow satisfy. So far, we have not yet found any other language that has quite the necessary set of features to provide an ergonomic mapping to HCL's infoset, beyond HCL's own native syntax (which was, of course, designed to be 1:1 with the infoset) and lower-level data structure representation languages like JSON by cheating and layering on the template syntax from the HCL native syntax.

    Because the OpenTofu runtime is tightly coupled with the OpenTofu language, and the OpenTofu language is tightly coupled with HCL's infoset, Introducing a new language syntax into OpenTofu is most likely to involve mapping that new language's concepts onto the HCL infoset, and so we'd need to find a language where such a mapping is possible and ergonomic. We have not yet identified any such language.

- Mapping a higher-level language onto a lower-level one (or sometimes, the opposite!) is commonly achieved using some form of preprocessing, such as transpilation or compilation.


    That technique has been employed successfully in several ecosystems -- including ours -- though of course it does come at an additional overall complexity cost that some may find unfavorable. I think preprocessing-like techniques are the most likely to succeed in the goal of bringing other languages into OpenTofu.

- It is a valid observation that there are both advantages and disadvantages in having a tightly-coupled domain specific language, and it's valid and expected for different people to have different preferences in this area.

    However, OpenTofu was designed for the domain-specific-language side of this tradeoff, and Pulumi already exists to satisfy the other end of that spectrum. I would suggest that those who prefer to work in domain-specific languages choose to use Pulumi instead, because it already has a successful design and implementation for that model of work.

---

To sum up, then: **I don't think this _specific_ request can really go anywhere** unless we can identify a single language that meets the difficult somewhat-conflicting requirements of being both widespread enough to already have good IDE/editor support _and_ adaptable enough to have HCL's infoset (and thus, the OpenTofu language) mapped to it ergonomically.

We could potentially evaluate ways in which OpenTofu could better support preprocessor-like solutions -- which could involve both changing the OpenTofu language to be easier to generate programmatically, or offering some way to invoke a preprocessor from within OpenTofu itself, or something else -- but we should probably treat those concerns as separate feature requests rather than trying to overload this issue to represent all of these possibilities at once. (If you _are_ a preprocessor author that has some feedback about how your preprocessor and OpenTofu are currently interacting, please open a new issue to discuss that!)

Therefore I'm going to mark this issue as blocked for now, since I don't expect we'll be able to do anything with it until we identify a suitable language. I'm also going to change the issue summary slightly since I think we've already ruled out Starlark as being suitable. New comments are welcome to identify candidate languages for us to consider, but for now this feature request is not directly actionable.

Thanks!

**Comment by nikolay** [comment]
@apparentlymart Thank you very much for the extensive analysis! I want to add just to other notes:

* C/C++ come with macros
* Rust has a powerful macro system
* Zig has an amazing comptime feature, which made many people leave Go for Zig.

The concept of having compile time and runtime feature split is a common pattern, which currently HCL does not support and that allowed several non-standard products to be created.

Just like HCL is a superset of JSON, maybe OpenTofu could fork HCL, too, which is stale in the hands of HashiCorp, and extend it with macros, etc., making it more appealing to the world outside of OpenTofu. This way, Terraform files will still work with OpenTofu, but if you write something specifically for OpenTofu, it could use extended syntax.

Many of the proposed features such as reusable types and even interpreters could also simply be accomplished with macros/preprocessor without significant changes to OpenTofu.

**Comment by bioball** [comment]
We'd love Pkl to be considered for this. It's designed specifically for configuration; we want our users to be able to manage complexity in config easily, and for it to be maintainable at scale. Also, we prioritize excellent tooling (e.g. IDE plugins, package sharing, in-language test framework).

Some notes about your requirements:

> It must be possible to enumerate the top-level declarations before fully evaluating expressions inside them, so that OpenTofu can build its dependency graph and evaluate objects in the correct order.
>
> For example, OpenTofu needs to be able to discover that there is something equivalent to a resource "aws_instance" "example" block before evaluating any expressions that might include a reference to that block and before evaluating the contents of that block.

The Pkl language is lazy; it only executes expressions that are in the critical path of evaluation, and you can control what that critical path is. This means that you can evaluate top-level properties (e.g. give me all resource names) without evaluating anything else.

> There must be something comparable to "unknown values" that can serve as placeholders in expressions to allow OpenTofu to produce partial results during the planning phase and then re-evaluate with more information during the apply phase.
>
> This one is unfortunately the trickiest to match in other languages: the type system under HCL has this feature largely because software like OpenTofu needs it, whereas more general-purpose languages don't tend to assume such a constraint.
>
> Pulumi emulates this using a promise-like model, which means that [a saved plan can fail during apply if the program behaves differently once values become "more known"](https://www.pulumi.com/blog/announcing-public-preview-update-plans/). OpenTofu's execution model instead assumes that such differences ought to be impossible at the language level and so always blames any inconsistencies on the providers generating the planned actions, since that's where the imperative code rules live in OpenTofu.

We don't have a solution for this today, but we want to support this in the future. See https://github.com/apple/pkl/issues/912 for reference.

**Comment by leonardo-zup** [comment]
Why not use a well-known scripting language like Lua to generate the `tf.json`? It seems like the simplest path to replace HCL while gaining all the flexibility and power of Lua.
